File: src/#515/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

var largestValues = function (root) {
  let res = [];

  helper(root, 0);

  return res;

  function helper(root, level) {
    if (root === null) {
      return;
    }

    if (level >= res.length) {
      res.push(root.val);
    } else if (root.val > res[level]) {
      res[level] = root.val;
    }
    helper(root.left, level + 1);
    helper(root.right, level + 1);
  }
};

File: src/#515/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func largestValues(root *TreeNode) []int {
	res := []int{}
	helper(root, 0, &res)
	return res
}

func helper(root *TreeNode, level int, levelList *[]int) {
	if root == nil {
		return
	}

	if level >= len(*levelList) {
		*levelList = append(*levelList, root.Val)
	} else if root.Val > (*levelList)[level] {
		(*levelList)[level] = root.Val
	}
	helper(root.Left, level+1, levelList)
	helper(root.Right, level+1, levelList)
}
